### ISSUE TITLE
Error when stopping the server and external clients connecting

### ISSUE THREAD

**Issue by Antwns** [jchristn/WatsonTcp#99]
Hello! When using the `.Stop();` Command I get ": `Object` reference not set to an instance of an object.` I have defined my server to the top of my class as ``internal static WatsonTcpServer AntVaultServer = new WatsonTcpServer("192.168.1.121", Convert.ToInt32(AuxiliaryServerWorker.ReadFromConfig(true, false)));``

It grabs the port number from a config file I have created.
I don't think that this is an error on my code since I have even tried to define the server in a method that takes the server as a constructor, and I get the same error, I hope you can look into it when you have free time on your hands! 

EDIT: I also get this error when a computer tries to connect to my server: WatsonTCP.pdb could not be found

**Comment by jchristn** [comment]
Hi @Antwns would you be able to capture the full exception and paste it here?

**Comment by Antwns** [comment]
Here is the error regarding the ``.Stop();`` function.

![Screenshot_5](https://user-images.githubusercontent.com/38153661/96613960-e0a80400-12ff-11eb-83db-4fb7148d3680.png)
![Screenshot_1](https://user-images.githubusercontent.com/38153661/96615331-6bd5c980-1301-11eb-815f-19030fbd35d6.png)


And here is the error regarding the missing .pdb file

![download](https://user-images.githubusercontent.com/38153661/96614149-12b96600-1300-11eb-83c4-f8d47114bd09.png)
![Screenshot_1](https://user-images.githubusercontent.com/38153661/96614897-ea7e3700-1300-11eb-8cb1-1ffa9abbec8d.png)

This error only occurs when you try to connect to the server from a machine outside your own network, I made a hotspot with my phone and the clines just wouldn't connect, I would get an error on my console "Client [ip] disconnected due to normal".
Firewalls were disabled on both machines and a DMZ was set up for the host machine, tested and validated them with a game server to make sure that the ports are open and working. Machines outside the hosts' network just cannot connect due to the missing .pdb file.

I'm really thankful for all the hard work your putting in for all of us.
I'm fairly new to programming and have been working with C# for about 6 months, sorry if the information I provided were lacking

**Comment by jchristn** [comment]
Hi @Antwns thanks for the kind words.

I really, really hate to ask this, but would you be able to clone to your local machine, build in debug, then reference the debug DLL directly?  As far as I know there isn't a clean/easy way to include PDB files in NuGet packages.  If you reference the DLL locally, you could then put a try/catch around your call to ```.Stop()``` which would help identify the command happening inside that is causing the exception.

In the meantime I'll also put together a patch that will have an exception event in ```WatsonTcpServer.Events``` that will be invoked (if set) when an exception is generated.  If you prefer to wait for that, no problem.

**Comment by Antwns** [comment]
Hello @jchristn . Adding a try/catch does nothing, the program still breaks in debug mode. I tried adding the .pdb file after compiling the .dll file. The client from the external network still doesn't connect, it breaks with null reference on ``.Start();`` just like the ``.Stop();`` does. Here's the screenshot on more details on the exception.

![Screenshot_1 (2)](https://user-images.githubusercontent.com/38153661/96623547-d25fe500-130b-11eb-9750-75e41d04fa0a.png)

I hope it gets figured out soon! 🙏

**Comment by jchristn** [comment]
Sounds like there are two issues:

1) external clients not able to connect
and
2) ```Stop()``` throwing a mysterious ```NullReferenceException```

For (1), internal clients are able to connect ok?  From the external machine, can you successfully ```telnet``` to the port on the server?  i.e. ```telnet [public ip address]:[port]``` for example ```telnet 73.93.x.x 8000```?

For (2) I updated NuGet with a patch version 4.4.0.2 that includes a new event method ```WatsonTcpServer.Events.ExceptionEncountered```.  Could you implement a handler for this event and send me the ```Json``` property from the ```ExceptionEventArgs``` that you receive?  You could also assign ```WatsonTcpServer.Settings.Logger = Debug.WriteLine``` should should give you similar output.

**Comment by Antwns** [comment]
1)I have setup a DMZ server and have verified that the ports are open. External clients still cannot connect.
2)Local clients now cannot connect either for some reason(I really don't know, I didn't change anything in my code) with timeout exception.
3)I cannot do step two because local clients cannot connect either now... I just get a timeout exception

**Comment by Antwns** [comment]
@jchristn I just figured out that there is an issue with my router, I'm looking into it right now! SORRY

**Comment by Antwns** [comment]
Ok so @jchristn , local clients can connect normally, no problem. Remote clients get Null reference exception for some reason now only.

**Comment by jchristn** [comment]
Ok - thanks.  Do they only get the exception after you've stopped the server?  Or even when the server is started?

**Comment by Antwns** [comment]
I see null reference exception when: 
1)Connecting to the server from a machine outside the local network.
2)Stopping the server after the .Start(); Command. 
I have defined .KeepAlive settings identical to the ones stated on the Readme.MD file. I've also only setup MessageDataReceived on my client and server, hope this info is useful as well

**Comment by jchristn** [comment]
For (1), do you see that exception on the client?  

For (2), you see this on the server?

**Comment by Antwns** [comment]
They both break the program in debug mode, even if I try try/catch, it still breaks. I don't get any details on the exceptions other than the ones I provided you with... I'm also using the Nuget pack with WPF and not Winforms, is that bad?

**Comment by jchristn** [comment]
No, not bad at all :) 

You mention above that you are instantiating using ```internal static WatsonTcpServer AntVaultServer = new WatsonTcpServer("192.168.1.121", Convert.ToInt32(AuxiliaryServerWorker.ReadFromConfig(true, false)));``` 

Is this inside of a form/view by chance?

I tested using a VM on EC2 and my local machine and can't reproduce.

**Comment by Antwns** [comment]
No this code is in it's own class. I don't use winform hosts either.
There's a function doing ``StartAntVaultServer();`` which is as follows:
``
        public static void StartAntVaultServer()
        {
            if (SetUpEvents == false)
            {
                AntVaultServer.Keepalive.EnableTcpKeepAlives = true;
                AntVaultServer.Keepalive.TcpKeepAliveInterval = 5;
                AntVaultServer.Keepalive.TcpKeepAliveTime = 5;
                AntVaultServer.Keepalive.TcpKeepAliveRetryCount = 5;
                AntVaultServer.Events.ClientConnected += Events_ClientConnected;
                AntVaultServer.Events.ClientDisconnected += Events_ClientDisconnected;
                AntVaultServer.Events.MessageReceived += Events_DataReceived;
                SetUpEvents = true;
            }
            CheckConfig(AuxiliaryServerWorker.ConfigFileDir);
            CheckUserDatabase(AuxiliaryServerWorker.UserDatabaseDir);
            CheckFileFolders();
            AntVaultServer.Start();
            AuxiliaryServerWorker.WriteToConsole("[INFO] Server started on port " + AuxiliaryServerWorker.ReadFromConfig(true,false));
        }
``

**Comment by jchristn** [comment]
Ok good to know.

Could you try in a non-static class?  My worry is that the WatsonTcpServer object is being garbage collected since the ```StartAntVaultServer``` method will exit.  You'll notice in each of the test projects (for instance, Test.Server: https://github.com/jchristn/WatsonTcp/tree/master/Test.Server) there is a while loop which keeps the WatsonTcpServer object from being garbage collected until exit.

You could also try something like this in the static class, after you've finished all of your other code:
```
using System.Threading;
...
EventWaitHandle waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset, null);
bool waitHandleSignal = false;
do
{
    waitHandleSignal = waitHandle.WaitOne(1000);
}
while (!waitHandleSignal);
```

**Comment by Antwns** [comment]
I could try but how would I call the start function since it's in another class file? 😢

**Comment by jchristn** [comment]
If you change your ```StartAntVaultServer``` method to the code above you won't need to put it into an instance.
```
using System.Threading;
...
public static void StartAntVaultServer() 
{
    if (SetUpEvents == false) 
    {
        AntVaultServer.Keepalive.EnableTcpKeepAlives = true;
        AntVaultServer.Keepalive.TcpKeepAliveInterval = 5;
        AntVaultServer.Keepalive.TcpKeepAliveTime = 5;
        AntVaultServer.Keepalive.TcpKeepAliveRetryCount = 5;
        AntVaultServer.Events.ClientConnected += Events_ClientConnected;
        AntVaultServer.Events.ClientDisconnected += Events_ClientDisconnected;
        AntVaultServer.Events.MessageReceived += Events_DataReceived;
        SetUpEvents = true;
    }

    CheckConfig(AuxiliaryServerWorker.ConfigFileDir);
    CheckUserDatabase(AuxiliaryServerWorker.UserDatabaseDir);
    CheckFileFolders();
    AntVaultServer.Start();
    AuxiliaryServerWorker.WriteToConsole("[INFO] Server started on port " + AuxiliaryServerWorker.ReadFromConfig(true, false));
    
    EventWaitHandle waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset, null);
    bool waitHandleSignal = false;
    do
    {
        waitHandleSignal = waitHandle.WaitOne(1000);
    }
    while (!waitHandleSignal);
}
```

Also could you set the ```WatsonTcpServer.Settings.Logger = Debug.WriteLine``` and ```WatsonTcpClient.Settings.Logger = Debug.WriteLine``` and capture the output?  That will show if client or server are being disposed (could be initiated by garbage collection).

**Comment by Antwns** [comment]
I'll do it and I'll let you know about it as soon as possible! 
PS: Adding 
`
EventWaitHandle waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset, null);
`
`bool waitHandleSignal = false;`
`do`
`{`
     `waitHandleSignal = waitHandle.WaitOne(1000);`
`}`
`while (!waitHandleSignal);`
`

Makes the app unresponsive because it's being handled by the main thread(STA)

**Comment by jchristn** [comment]
Got it.

I would definitely recommend creating an instance class then where you have WatsonTcpServer and WatsonTcpClient.  It's generally not a good practice to put socket/external communications libraries in the main thread or in a form directly for that matter.  Things get garbage collected, main thread gets blocked, etc.  There should be an ```App.xaml.cs``` or similar where you could instantiate your class containing WatsonTcpServer.

**Comment by Antwns** [comment]
Hi. I tried moving the server start function to a different thread with
       ``private void ServerStartButton_Click(object sender, RoutedEventArgs e)``'
       ``{``
           ``Thread ServerStartThread = new Thread(()=>MainServerWorker.StartAntVaultServer());``
           ``ServerStartThread.IsBackground = true;``
            ``ServerStartThread.Start();``
            ``ServerStartButton.IsEnabled = false;``
            ``ServerStopButton.IsEnabled = true;``
         ``}``

I'm sorry, I'm pretty new to programming as I said and I'm learning on my own

**Comment by jchristn** [comment]
Good to know.  

Regarding the exception, did you implement ```.Settings.Logger``` on both client and server?  Could you send me the output?

Are the outside hosts able to connect?

**Comment by Antwns** [comment]
I do not know how to enable the logger, could you give me an example? 
Outside clients are still not able to connect although I no longer get null reference exception, they just disconnect immediately for no reason.
Maybe if I can setup the logger we can figure it out!

**Comment by jchristn** [comment]
You could set up a method like this:
```
using System.IO;
...
server.Settings.Logger = WriteToLog;

public void WriteToLog(string msg)
{
  File.AppendAllText(@"log.txt", msg + Environment.NewLine);
}
```

Do this on both client and server and paste the contents of each file here.  Cheers!

**Comment by Antwns** [comment]
Hello! Here is the log results from the server:
`[WatsonTcpServer] starting on 192.168.1.121:8910
[INFO] Server started on port 8910
[WatsonTcpServer] accepted connection from 151.68.174.247:20366
[WatsonTcpServer] starting data receiver for 151.68.174.247:20366
[WatsonTcpServer] listener stopped
[INFO] Received new connection from 151.68.174.247:20366
[INFO] Received new connection from 151.68.174.247:20366
[WatsonTcpServer] message build failed due to disconnect for client 151.68.174.247:20366
[WatsonTcpServer] data receiver terminated for 151.68.174.247:20366
[INFO] Client 151.68.174.247:20366 disconnected due to Normal
[INFO] Client 151.68.174.247:20366 disconnected due to Normal
[WatsonTcpServer] disposing data receiver for 151.68.174.247:20366
`
And from the client:
`[WatsonTcpClient] message build failed due to disconnect[WatsonTcpClient] data receiver terminated[WatsonTcpClient] disposing[WatsonTcpClient] disposed`

It doesn't tell me much...

**Comment by jchristn** [comment]
Well, you just discovered a bug :)  I had a ```?``` incorrectly placed in the ```Stop()``` method.  Going to release a patch on that right now.

The client logs indicate that something is interrupting the client connection.  It also appears the client is connecting multiple times or somehow the server's ```ClientConnected``` event is firing more than once.  I assume this output is coming from your ```ClientConnected``` event:
```
[INFO] Received new connection from 151.68.174.247:20366 
[INFO] Received new connection from 151.68.174.247:20366 
```
Are you sure you're instantiating the client only one time?

Something is definitely interrupting the client connection:
```
[WatsonTcpClient] message build failed due to disconnect
[WatsonTcpClient] data receiver terminated
[WatsonTcpClient] disposing[WatsonTcpClient] disposed
```

Could you make one more change?
```
client.Events.ExceptionEncountered += ExceptionEncountered;

public void ExceptionEncountered(object sender, ExceptionEventArgs args)
{ 
  File.AppendAllText(@"log.txt", args.Json + Environment.NewLine);
}
```

And send me that output?  You can do the same on the server side as well.

**Comment by Antwns** [comment]
Yeah I made my own logger long before I started using WatsonTCP 😄 
I get these errors on the client side:

{
  "ClassName": "System.TimeoutException",
  "Message": "Timeout connecting to AntMax.ddns.net:8910",
  "Data": null,
  "InnerException": null,
  "HelpURL": null,
  "StackTraceString": "   at WatsonTcp.WatsonTcpClient.Start()",
  "RemoteStackTraceString": null,
  "RemoteStackIndex": 0,
  "ExceptionMethod": "8\nStart\nWatsonTcp, Version=4.4.0.3, Culture=neutral, PublicKeyToken=null\nWatsonTcp.WatsonTcpClient\nVoid Start()",
  "HResult": -2146233083,
  "Source": "WatsonTcp",
  "WatsonBuckets": null
}
On the server side I have no error log generated.

**Comment by jchristn** [comment]
Thanks.  Something's blocking or the listener isn't set up properly given your NAT configuration.  I assume you have a router/firewall rule that forwards port 8910 to your machine's internal IP address?  Are you running any other services on that machine (for instance a webserver) where something is able to successfully get through?  Have you confirmed that the IP address reported by ping (I tested it and it came back as 151.68.174.247) matches what you see on a site like www.whatismyipaddress.com?

**Comment by Antwns** [comment]
My router is experiencing issues today, and I don't know why. As soon as I have more info I will get back to you about it, I have setup a DMZ on it but for some reason it stops working after a while...

**Comment by jchristn** [comment]
Sounds good @Antwns let me know :)

**Comment by Antwns** [comment]
I discovered that as soon as I launch the server my whole NAT crashes and no new connections would be accepted , I reset my router about 10 times to reach that conclusion. I really don't know what's going on...

**Comment by Antwns** [comment]
Hi @jchristn , I just had an idea, could it be possible that the client also needs port forwarding and not just the server to use this particular package? I've been banging my head around this

**Comment by jchristn** [comment]
Hi @Antwns client-side NAT almost universally always automatically does this.  Otherwise the Internet would not because the return trip from the server to respond to the client would be broken :)

I'll take a look today and believe me I won't judge or be triggered!

**Comment by Antwns** [comment]
Thank you @jchristn you're a true role model for aspiring programmers!

**Comment by jchristn** [comment]
Hi @Antwns something to consider - I'd recommend two things as I'm starting to go through your server code:

1) I'd avoid having any kind of cyclical classes, i.e. you have MainServerWorker calling AuxiliaryServerWorker and vice versa
2) I'd recommend consolidated these two into an instance class, defining the class in App.xaml.cs, and instantiate in the main window

I'm going to make some suggested changes in the code based on what I'm seeing and post it here.  Though I hate to ask you to undergo a major refactor, I think you'll end up in a better place in the long run, and it will make it easier for me to debug :)

**Comment by Antwns** [comment]
Hi @jchristn I'm always willing to learn how to do better! I might even restart my project again with the chances you'll recommend, thank you so much for taking time out of your day to look into my code!

**Comment by jchristn** [comment]
Happy to.  Another modification:
```
if (str.StartsWith("/NewConnection"))//NewConnection -U Username -P Password.
{
    Thread AuthenticationThread = new Thread(() => DoAuthentication(str, e));
    AuthenticationThread.Start();
}
```
Could be:
```
Task.Run(() => DoAuthentication(str, e));
```
You could further add a ```CancellationToken``` which would allow you to cancel the tasks later if you want.

**Comment by jchristn** [comment]
Hi @Antwns I can't run your code, and looking at the original source I can't point to something keeping it from working.  I did however put in a minor refactor, if you'd like to try it.  I didn't make any changes to the client, only to the server, to instantiate the MainServerWorker off of App.xaml.cs.  If you update your files with the copy attached here, you can remove the AuxiliaryServerWorker.cs file altogether, and the server should run as a background task and off of the main thread. 

I just tested again connectivity between WatsonTcp client and server between a VM on Azure and a VM on EC2 and it works fine.  I suspect there's a networking problem somewhere on your side.

Note that I haven't tested any of these code changes, so please make a backup in case I've broken it terribly.  I used the typical styling that I use in my projects from a public vs private variable naming perspective and organized into sensible regions.

Cheers, Joel 

[patch.zip](https://github.com/jchristn/WatsonTcp/files/5424294/patch.zip)

**Comment by jchristn** [comment]
Hi @Antwns any update on your side?  Cheers, Joel

**Comment by Antwns** [comment]
@jchristn Hello, I was dealing with some health issues the past few days, I'll be giving it a shot tomorrow first hand and I'll let you know what happens. It COULD be that my router is just faulty, however other applications work just fine and only this package presents those problems... Like I said, I'll be giving you more definitive details tomorrow. Thank you so much for taking time out of your day to look into my code again, I really appreciate it! 😄

**Comment by jchristn** [comment]
Hope you're feeling better @Antwns - I also published a new version with some other related fixes. Cheers

**Comment by Antwns** [comment]
Hello @jchristn , applying your patch unfortunately did not fix the issue, however I do have a good idea as of what could be causing this. I will 
1)Switch to a console based server instead of WPF to avoid the weird threading stuff you mentioned above.
2)When I created the projected I checked the box in visual studio to place the project files in the same folder(Or something similar to that) I think it's messing with the assembly names or something. I will recreate the entire project and import the files as well as making the changes mentioned above. I'll post the results here when I'm done making these changes, cheers! 😄
